### ISSUE TITLE
knex.val() similar to knex.ref

### ISSUE THREAD

**Issue by Gregroam** [tgriesser/knex#2823]
# Feature discussion / request

Is there something like `knex.val('hello world')` similar to `knex.ref('column name')`? Use case would be something like join clauses.

```
knex("foo").select().leftJoin("bar", "column", knex.val('hello world'))
```

this is just a simplified example, there are cases when you need to compare a column to a value in a join condition, but knex assumes the left and right hand sides of the expression will be column names. Currently the workaround is just to use
`knex.raw('?', 'hello world')`, but since you've added knex.ref which is an alias `knex.raw('??', 'column')`, would we be able to get something similar for values, if there isn't already something for that?

**Comment by wubzz** [comment]
If this is just a wrapper around `knex.raw('?', value)` then it's very simple to add in the same was `.ref` was added.

If there are additional features for it to handle lets document them here before starting to work on it.

**Comment by tgriesser** [comment]
Actually, I think we should add `leftJoinVal` and `rightJoinVal` to match the `onVal` syntax that was merged, since it's a common enough use case.

**Comment by tgriesser** [comment]
Also, once we have that I think I'd be good for a 0.16 release

**Comment by Gregroam** [comment]
What if you want to mix columns and values in a single join object?

```
.leftJoin({
  "tableA.colA": "tableB.colA",
  "tableB.colB": knex.val(true)
})
```

That would be nice to have

**Comment by tgriesser** [comment]
Fair enough. I'd be fine accepting a PR for `knex.val`.

**Comment by wubzz** [comment]
@tgriesser Simple `knex.raw('?', value)` proxy or do we want it to be more sophisticated than that?

In theory this could be used for postgres arrays where we currently have to stringify the input. IE
```javascript
knex('test').insert({list: ['1', '2', '3']}).toString();
//insert into "test" ("list") values ('{"1","2","3"}')

knex('test').insert({list: JSON.stringify(['1', '2', '3'])}).toString();
//insert into "test" ("list") values ('["1","2","3"]')

knex('test').insert({list: knex.val(['1', '2', '3'])});
//insert into "test" ("list") values ('["1","2","3"]')
```

That is to say `knex.val` would stringify it automatically.

Personally I think this would just make things a bit more confusing, so I'd opt for a simple `Raw` proxy the same way `ref` works. Is that fine?

**Comment by Gregroam** [comment]
Maybe a way to cast its type?
```
knex.select(knex.val(`{"a": 1, "b": 2}`).cast("json"))
```

produces in postgres
```
SELECT '{"a": 1, "b": 2}'::json
```

**Comment by elhigu** [comment]
@Gregroam: @wubzz had IMO a better idea how to do the casting in a way that it could be use in more places.

**Comment by wubzz** [comment]
Cast proposal is #2587 - In this scenario it would be
```javascript
knex.select(knex.castJson(knex.val(''{"a": 1, "b": 2}'')))
```
_(Actually would work even without knex.val in that example)_

**Comment by Gregroam** [comment]
Oh nice. Yeah that sounds good. Can't think of any other special needs that knex.val would need. Just make it a wrapper for `knex.raw('?', binding)`

**Comment by mercmobily** [comment]
Currently, when you need to use a value (not a column reference) in certain contexts like join conditions, you have to use the workaround:
```javascript
knex.raw('?', 'hello world')
```

This is inconsistent with the API since we have `knex.ref()` for column references, which is essentially:
```javascript
knex.raw('??', 'column_name')
```

**Proposed Implementation**

Based on the discussion, `knex.val()` should be a simple wrapper around `knex.raw('?', value)`, similar to how `knex.ref()` works. This would provide:

1. **API Consistency**: Matching pair with `knex.ref()`
2. **Clearer Intent**: Makes code more readable by explicitly indicating a value vs a column reference
3. **Better Join Support**: Would work nicely with the `onVal` methods that were added

Example usage:
```javascript
// In joins
knex("foo").leftJoin("bar", function() {
  this.on("bar.status", "=", knex.val('active'))
})

// Mixed columns and values
.leftJoin({
  "tableA.colA": "tableB.colA",
  "tableB.active": knex.val(true)
})
```

**Implementation Details**

The implementation would be straightforward, following the same pattern as `knex.ref()`:

```javascript
// In the Knex client
val(value) {
  return this.raw('?', value);
}
```

**Why It Hasn't Been Done**

Despite @tgriesser saying he'd "be fine accepting a PR for `knex.val`" and the general agreement on its usefulness, it appears no one has submitted a PR yet. This would be a good first contribution as it's:
- Simple to implement
- Has clear requirements
- Follows existing patterns
- Has consensus from maintainers

**Alternative Approaches Mentioned**

@tgriesser also suggested adding `leftJoinVal` and `rightJoinVal` methods to match the `onVal` syntax, though the consensus seemed to favour the more general `knex.val()` approach as it's more flexible.

**Related Work**

- `onVal` methods were added for value comparisons in join clauses (#2746)
- Cast proposal (#2587) would work well with `knex.val()`
- `knex.ref()` provides the pattern to follow

**Recommendation**

This feature should be implemented as it would improve API consistency and developer experience. The implementation is straightforward and has maintainer approval.

**Related issues**

- #2746 - Added onVal methods for join clauses
- #2587 - Cast proposal that would complement knex.val()
- #2831 - Related issue about values in join conditions
